decentraland/unity-explorer
# fix: sdk hot reload content mappings update + landscape

## WHY

During local scene development, if the locally running scene updates its mappings (for example if a new GLTF 3D model asset is added in any of its directories), the Explorer reloads the scene but doesn't fetch the `SceneEntityDefinition` from scratch, thus keeping an old version of the content/mappings available in the scene.

This is something used VERY frequently with Decentraland Editor (the level editor accessible through the new Creators Hub) and it's very problematic that the Explorer is not able to reload correctly when a new 3D asset is placed in the scene.

## WHAT

Updated the scene reload logic, only for local scene development, to make the pertinent systems fetch everything scene-related from scratch when reloading during hot-reload.

#### Bonus

Since having the local scene as a static/predefined scene makes it a lot easier, enabled the Landscape with dynamic generation for local scene development as well.


## QA TEST INSTRUCTIONS

1. Clone/Download the [sdk7 template scene](https://github.com/decentraland/sdk7-scene-template)
2. open the file `index.ts` and **replace all the code in that file** by the following snippet:

```
import { Vector3 } from '@dcl/sdk/math'
import { engine, Transform, GltfContainer, MeshRenderer } from '@dcl/sdk/ecs'
const sphereEntity = engine.addEntity()
Transform.create(sphereEntity, { position: Vector3.create(8, 1, 8) })
MeshRenderer.setSphere(sphereEntity)
const gltfEntity = engine.addEntity()
Transform.create(gltfEntity, { position: Vector3.create(8, 3, 8) })
GltfContainer.create(gltfEntity, { src: 'shark.glb' })
```

3. Install the SDK's `latest` package running `npm i @dcl/sdk@next` in the root folder of the scene.
4. Run `npm run start -- --explorer-alpha` to start the scene
5. Kill the launcher/explorer that will auto-open, but leave the server running in the console
6. Download the build from this PR
7. Open the build from this PR connecting it to the locally running scene [WIKI steps for that](https://github.com/decentraland/unity-explorer/wiki/How-to-connect-to-a-local-scene-(Unity-Editor---Custom-Build---Latest-Released-Build)#connecting-a-custom-build-to-the-scene) (remember to put the correct scene position like: `position 0,0`)
8. When the Explorer loads the local scene: **confirm that the landscape is enabled and it is limited around the scene as if it were a world. But don't get too far from the scene as that triggers a known issue**
9. Also confirm that there is a sphere object in the middle of the scene
10. Step into the scene boundaries and keep the scene running in the Explorer and download [this 3D model (shark.glb)](https://github.com/decentraland/sdk7-test-scenes/raw/129a4b1553d5c340d52c1e4090a9e9a1063b47c9/scenes/73,-8-animator-tests-shark/models/shark.glb) and put it in the root folder of the scene
11. Confirm that the scene gets automatically reloaded in the Explorer and that now you see a Shark 3D model on top of the sphere

**NOTE (Known issue): If you get more than 1 parcel away from the scene, the hot reload will stop working until you get on the scene again and run the `/reload` chat command**. Dismiss [the known issue](https://github.com/decentraland/unity-explorer/issues/2357) as it will be tackled separately.

https://github.com/user-attachments/assets/ae6d0c46-2902-4b2a-98d4-dd4a6965b0e7
